sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala

```diff
@@ -261,6 +261,20 @@ trait ExternalCatalog {
       table: String,
       partialSpec: Option[TablePartitionSpec] = None): Seq[CatalogTablePartition]
```

Comment:
@beliefer should we instead add a pattern to `listPartitions`?

Reply:
LGTM if add a pattern to `listPartitions`.

Reply:
After a second thought. Take 'a=1/b=1' as an example, the pattern string hard to describe partial partition spec.

Reply:
Yeah, it's hard to add pattern to `listPartitions`.  `getPartitionsByNames` also had argument with `Map<String, String> partialPartSpec` like `getPartitionsByNames(Table tbl,
      Map<String, String> partialPartSpec)`. It's more efficient in Hive. However, it still pull all partitions info when using dynamic partition overwrite.